Cache `/teachers/classrooms/classrooms_i_teach`

anathomical opened the pull request:
## WHAT
Cache `/teachers/classrooms/classrooms_i_teach` payload
## WHY
In our effort to cache all of our reporting endpoints to reduce db load
## HOW
Wrap the code that generates the payload in a cache call at the all classrooms level

### Notion Card Links
https://www.notion.so/quill/Turn-Caching-Back-On-4bf7669d517443909640387fbafdc958

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Deploying now
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Cache `/teachers/classrooms/classrooms_i_teach`

anathomical committed:
Add test for cache behavior

anathomical committed:
Refactor caching to allow for more robust tests

anathomical committed:
Fix linting error

brendanshean commented on services/QuillLMS/app/controllers/teachers/classrooms_controller.rb:
```diff
@@ -158,6 +157,13 @@ def transfer_ownership
+  private def fetch_classrooms_i_teach_cache
+    current_user.all_classrooms_cache(key: 'teachers.classrooms.classrooms_i_teach') do
+      @classrooms = current_user.classrooms_i_teach
+      @classrooms.sort_by { |c| c[:update_at] }
```
I'm guessing this is an older bug, but should this be `:updated_at`?

anathomical replied:
Yeah, you can see that this code used to live on line 35, and I moved it.  While this probably should be fixed, I'm hesitant to actually do so in case there are now expectations on the current (probably broken) behavior.  Thoughts?

brendanshean replied:
That's a good point. What about removing the sort_by altogether.

brendanshean replied:
The [docs]( https://ruby-doc.org/core-2.7.0/Enumerable.html#method-i-sort_by) mention that `sort_by` maps each value in the array to the block, but if the mappings are equal, the result is unpredictable.  Trying it out in the console a few times seems to return the original array.

anathomical replied:
Seems reasonable.  I'll pull the sort and make sure tests pass, and we can cross our fingers that there's no weird gotchas in there.

brendanshean commented on services/QuillLMS/spec/controllers/teachers/classrooms_controller_spec.rb:
```diff
@@ -316,14 +316,25 @@
 
     before do
-      allow(teacher).to receive(:classrooms_i_teach) { [classroom] }
+      allow(teacher).to receive(:classrooms_i_teach).once.and_return([classroom])
```
This is a very clever approach to testing the caching.

brendanshean reviewed (approved):
There's a minor typo that should be fixed but nothing that should hold this up.  Nice work!

anathomical committed:
Remove sort that doesn't do anything

This code was copied from existing code, but since it seems extraneous (since it sorts on a non-existent key) we can just remove it while we're doing other things here
